[Backport release-24.05] keycloak: 24.0.5 -> 25.0.6

NickCao opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NickCao committed:
keycloak: 24.0.5 -> 25.0.1

Diff: https://github.com/keycloak/keycloak/compare/24.0.5...25.0.1
(cherry picked from commit 134a223f3fda26847e51cbe92db9b06a8142e322)

NickCao committed:
keycloak: enable hostname:v1 feature by default

teutat3s commented on pkgs/servers/keycloak/default.nix:
```diff
@@ -7,7 +7,7 @@
 , confFile ? null
 , plugins ? [ ]
-, extraFeatures ? [ ]
+, extraFeatures ? [ "hostname:v1" ]
```
Looks like this needs to be 
```suggestion
, extraFeatures ? [ "hostname-v1" ]
```

NickCao replied:
This fails even earlier, at the build time.

teutat3s replied:
It's failing [here](https://github.com/keycloak/keycloak/blob/25.0.1/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java#L310), might need to set `hostname-port`?

NickCao replied:
Doesn't work either.

teutat3s reviewed (commented):
It's probably failing because the feature name is not valid. I also saw references to `hostname:v1` in the docs, but it's seems the feature name is `hostname-v1`.
Source: https://www.keycloak.org/server/features

mweinelt commented:
Any particular reason for this backport?

teutat3s commented:
Some additional observations that caught my eye while trying to debug this. I could not spot clues to a solution yet.

```
❯ nix build --print-build-logs -f . nixosTests.keycloak.postgres.driverInteractive
...
WARN  [org.key.com.Profile] (main) Deprecated features enabled: hostname:v1
WARN  [org.key.qua.run.cli.Picocli] (main) The following run time non-cli options were found, but will be ignored during build time:
  kc.db-url-host, kc.db-url-database, kc.db-url-port, kc.db-url-properties, kc.db-username, kc.db-password, kc.hostname,
  kc.hostname-strict-backchannel, kc.http-host, kc.http-port, kc.https-port, kc.https-certificate-file, kc.https-certificate-key-file, kc.proxy
...
```
and
```
$ kc.sh --config-file /nix/store/j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf show-config
Current Mode: production
Current Configuration:
	kc.config.built =  true (SysPropConfigSource)
	kc.config.file =  /nix/store/j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf (SysPropConfigSource)
	kc.db =  postgres (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-password =  ******* (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-database =  keycloak (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-host =  localhost (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-port =  5432 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-properties =   (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-username =  keycloak (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.hostname =  acme.test (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-host =  0.0.0.0 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-port =  80 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-relative-path =  / (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-certificate-file =  /run/keycloak/ssl/ssl_cert (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-certificate-key-file =  /run/keycloak/ssl/ssl_key (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-port =  443 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.log-console-output =  default (classpath keycloak.conf)
	kc.proxy =  none (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.spi-hostname-v2-hostname =  acme.test (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.version =  25.0.1 (SysPropConfigSource)
```

r-ryantm committed:
keycloak: 25.0.1 -> 25.0.2

(cherry picked from commit 5182f25f5bbe33e936ed43ad04affd14e34c3698)

NickCao committed:
keycloak.plugins.keycloak-metrics-spi: 5.0.0 -> 6.0.0

Diff: https://github.com/aerogear/keycloak-metrics-spi/compare/refs/tags/5.0.0...6.0.0
(cherry picked from commit fc92a20d35d82be17863fbc7685c239f64d1949b)

r-ryantm committed:
keycloak: 25.0.2 -> 25.0.4

(cherry picked from commit 84c49afaaaa997662d819fbf998475803db4bbcf)

r-ryantm committed:
keycloak: 25.0.4 -> 25.0.5

(cherry picked from commit c6fa7e48fee0dbd532213eb34e972286bc828f7c)

NickCao committed:
keycloak: 25.0.5 -> 25.0.6

Diff: https://github.com/keycloak/keycloak/compare/None...25.0.6
(cherry picked from commit fcb30ff38b578ab4b19271de14f6acb5c63d5bb4)

NickCao committed:
nixos/keycloak: link $out/lib to KC_HOME_DIR to fix loading optimized app image

Reference: https://www.keycloak.org/docs/latest/upgrading/index.html#additional-validations-on-the-optimized-startup-option

NickCao commented:
In an unexpected turn of events, I found out why this was not working in https://github.com/NixOS/nixpkgs/pull/346439

NickCao committed:
nixos/keycloak: raise memory size in tests to 2047M

(cherry picked from commit 937d3964c905ce30cf227da1ccc826b6821214cc)

teutat3s commented:
Result of `nixpkgs-review pr 323480` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 tests built:</summary>
  <ul>
    <li>nixosTests.keycloak.mariadb</li>
    <li>nixosTests.keycloak.mysql</li>
    <li>nixosTests.keycloak.postgres</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pkgs.keycloak</li>
  </ul>
</details>

teutat3s commented:
@ofborg test keycloak

NickCao commented:
Let's add the commit raising test vm memory size, should help with the flaky tests.

NickCao commented:
So they only support one major version: https://endoflife.date/keycloak

We might as well leave it as is. Version 25 is no better than 24 in terms of support now.

dkowis commented:
https://www.keycloak.org/2024/10/release-updates

Just FYI, keycloak is changing the way they do releases so that it's less disruptive, and hopefully this will be easier to deal with.
